### ISSUE TITLE
Disallow values of type 'symbol' from being passed to 'Number' function/constructor

### ISSUE THREAD

**Issue by iansan5653** [microsoft/TypeScript#32895]
## Suggestion

Change the `NumberConstructor` interface to disallow passing values of type `Symbol`. Attempting to call `Number(Symbol())` or `new Number(Symbol())` both throw `TypeError`s.

## Use Cases

Any time a value of type `any` is passed to `Number`, there is a risk of that value being a `symbol` and throwing a `TypeError`.

### Checking if a value can be coerced to a number
I just ran into this issue when checking if an object key could be a valid array index. The following should be an error as it will throw if `value` is of type `symbol`.

```
function keyCanBeArrayIndex(value: string | number | symbol): boolean {
  return !Number.isNaN(Number(value));
}
```

Having this be invalid would have forced me to write something like:

```
function canBeArrayIndex(value: string | number | symbol): boolean {
  try {
    return !Number.isNaN(Number(value));
  } catch {
    return false;
  }
}
```

## Checklist

My suggestion meets these guidelines:

* [x] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, etc.)
* [x] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

**Comment by iansan5653** [comment]
Something like this would work, but it has a pretty unhelpful error text:
```ts
type NotSymbol<T> = T extends symbol ? never : T;
interface NumberConstructor {
    new<T>(value?: NotSymbol<T>): Number;
    <T>(value?: NotSymbol<T>): number;
    //...
}
```

**Comment by DanielRosenwasser** [comment]
Do you happen to pass `symbol`s into the `Number` function a lot? In my experience these sorts of changes don't give great returns for the amount of breakage they cause.

**Comment by iansan5653** [comment]
@DanielRosenwasser not a lot, no. However it certainly can happen. Personally I ran into the above example when making a `Proxy` to wrap an array. I wanted to trap the `set` action but only if the key was numeric. For most keys out there today the function would work perfectly, but in a couple years as symbols take off more and more, the program may start crashing.

Yeah it's not an issue you run into every day, but that's the whole benefit of using TypeScript anyway - it's supposed to prevent you from making these kind of mistakes without realizing it. 

---

Also, it's a really unintuitive behavior. When we do `` `${Symbol()}` ``, TS errors with:
```
Implicit conversion of a 'symbol' to a 'string' will fail at runtime. Consider wrapping this expression in 'String(...)'.
```
First of all that means TS is already attempting to prevent cases like these, and second of all it explicitly says that `String(Symbol())` is perfectly valid, which is true.  But I, like most people I think, would also assume that `Number(Symbol())` would be valid too, especially considering that with _any other value_, `Number` will yield a number or `NaN`.

**Comment by fatcerberus** [comment]
I think part of the problem here is IIRC TypeScript doesn't have a way to express the type constraint of "anything *except for* `symbol`"  Negated types would allow that, but we don't have them right now (see #29317).  In my experience people *do* use the `Number()` constructor for all kinds of runtime conversions, so overconstraining its argument to, e.g. `number | string` will most likely break real-world code.

**Comment by iansan5653** [comment]
>  In my experience people do use the `Number()` constructor for all kinds of runtime conversions

This is exactly why we should error for this case - people are using `Number()` for runtime conversions and I'm certain that most people have no idea that passing a symbol would throw an error, crashing their program. 

I completely agree that it should not be overconstrained. Unless there's something else that throws an error when passed to `Number` (like `{toString: null, valueOf: null}` does with `String`), _everything_ else should be valid.

**Comment by iansan5653** [comment]
Considering that #22251 was merged to prevent symbols from being used in template literals, this is clearly possible in some way and IMO should probably be handled the same way. So maybe we allow `String(any)` but not `String(symbol)`, just like that fix does. That behavior would handle my example case, which may (if I'm not mistaken) be the most common situation where you might encounter this:
![image](https://user-images.githubusercontent.com/2294248/63176953-e5469d80-c014-11e9-896d-fe8599848c49.png)

**Comment by iansan5653** [comment]
After some exploration, I found out that attempting to call `new String(Symbol())` (which is of course not a great idea anyway but it still is possible) also throws `TypeError`. I've updated the original issue to reflect that as well, since they are almost exactly the same situation.

In summary:
```ts
String(Symbol()); // fine; even recommended
new String(Symbol()); // TypeError

Number(Symbol()); // TypeError
new Number(Symbol()); // TypeError

Boolean(Symbol()); // fine
new Boolean(Symbol()); // fine

Object(Symbol()); // fine
new Object(Symbol()); // fine
```

**Comment by fatcerberus** [comment]
Any situation that requires a symbol to be coerced to a string (except for the special case of `String(symbol)`) will throw an error, per the standard.  There’s likely a lot of ground to cover here.

**Comment by iansan5653** [comment]
Yes, any time a symbol is implicitly coerced to a string throws an error, and that's exactly what should happen. That behavior is already very useful and works as expected.

The problem is, now TS throws an error for some cases of symbol type-conversion, but not others. We just need to handle these other cases as well.
